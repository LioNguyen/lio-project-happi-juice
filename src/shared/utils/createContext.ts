import type React from "react";
import { createContext as originCreateContext, useContext } from "react";

function createContext<T>(): [React.Provider<T>, () => T, React.Context<T>] {
  const context = originCreateContext<T>({} as T);

  return [context.Provider, () => useContext(context), context];
}

export { createContext };
