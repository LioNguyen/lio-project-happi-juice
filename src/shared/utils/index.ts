export * from './lazyLoad'
export * from './tailwind'
