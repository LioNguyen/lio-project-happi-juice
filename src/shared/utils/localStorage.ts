/**
 * Save data to localStorage
 * @param key - The key under which the data will be stored
 * @param value - The data to store, can be any serializable value
 */
function setLocalStorage<T>(key: string, value: T): void {
  try {
    const serializedValue =
      typeof value === 'string' ? value : JSON.stringify(value)

    localStorage.setItem(key, serializedValue)
  } catch (error) {
    console.error(`Error saving ${key} to localStorage:`, error)
  }
}

/**
 * Retrieve data from localStorage
 * @param key - The key of the stored data
 * @returns The parsed data or null if not found or error occurs
 */
function getLocalStorage<T>(key: string): T | null {
  try {
    const serializedValue = localStorage.getItem(key)

    if (!serializedValue || serializedValue === 'undefined') {
      return null
    }

    // Try to parse JSON
    try {
      return JSON.parse(serializedValue) as T
    } catch {
      // Return as string if JSON parsing fails
      return serializedValue as T
    }
  } catch (error) {
    console.error(`Error reading ${key} from localStorage:`, error)
    return null
  }
}

/**
 * Remove data from localStorage
 * @param key - The key to remove from localStorage
 */
function removeLocalStorage(key: string): void {
  try {
    localStorage.removeItem(key)
  } catch (error) {
    console.error(`Error removing ${key} from localStorage:`, error)
  }
}

/**
 * Clear all data from localStorage
 */
function clearLocalStorage(): void {
  try {
    localStorage.clear()
  } catch (error) {
    console.error('Error clearing localStorage:', error)
  }
}

export {
  clearLocalStorage,
  getLocalStorage,
  removeLocalStorage,
  setLocalStorage,
}
