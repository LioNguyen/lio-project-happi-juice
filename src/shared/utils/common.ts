import { format, isValid, Locale, parseISO } from 'date-fns'
import { enUS, vi } from 'date-fns/locale'
import _ from 'lodash'

import { IOrderItem } from '@/domains/order'
import { ISelectOption } from '@/shared/types'

interface INumberFormatOptions {
  decimals?: number
  thousandSeparator?: string
  decimalSeparator?: string
}

// ==================== Date Utilities ====================
/**
 * Parse various date formats into Date object
 * @category Date
 * @param {Date | string | undefined} value - Input date
 * @returns {Date | undefined} Parsed Date object or undefined if invalid
 * @example
 * parseDate("2024-02-26") // Date object
 * parseDate("invalid") // undefined
 */
function parseDate(value: Date | string | undefined): Date | undefined {
  if (!value) return undefined
  if (value instanceof Date) return isValid(value) ? value : undefined

  try {
    // Try ISO string parsing first
    const date = parseISO(value)
    if (isValid(date)) return date

    // Fallback to direct Date creation
    const normalDate = new Date(value)
    return isValid(normalDate) ? normalDate : undefined
  } catch {
    return undefined
  }
}

/**
 * Format a date string according to specified format
 * @category Date
 * @param {string} date - Date string to format
 * @param {string} [dateFormat] - Output format (defaults to "yyyy-MM-dd")
 * @returns {string} Formatted date string
 * @example
 * formatDateTime("2024-02-26") // "2024-02-26"
 * formatDateTime("2024-02-26", "dd/MM/yyyy") // "26/02/2024"
 */
function formatDateTime(date: string, dateFormat?: string) {
  return format(date, dateFormat || 'yyyy-MM-dd')
}

const localeMap: Record<string, Locale> = {
  vi: vi,
  en: enUS,
}

// ==================== Number & Price Utilities ====================
/**
 * Format a number with customizable separators and decimals
 * @category Number
 * @param {number} value - Number to format
 * @param {INumberFormatOptions} [options] - Formatting options
 * @returns {string} Formatted number string
 */
const formatNumber = (
  value: number,
  options?: INumberFormatOptions,
): string => {
  const {
    decimals = 0,
    thousandSeparator = ',',
    decimalSeparator = '.',
  } = options || {}

  const parts = value.toFixed(decimals).split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, thousandSeparator)

  return parts.join(decimalSeparator)
}

/**
 * Formats the price with thousand separators using locale-specific formatting
 * @category Price
 * @param {string} price - Price string to format
 * @returns {string} Formatted price string
 */
const formatPrice = (price: string) => {
  return Number(price).toLocaleString()
}

// ==================== Array Utilities ====================
/**
 * Transforms an array of select options into a sorted, joined string
 * @category Array
 * @param {ISelectOption[]} array - Array of select options
 * @param {string} [key] - Key to extract from each option
 * @param {string} [separator] - Separator for joining values
 * @returns {string} Sorted and joined string
 */
const transformToSortedString = (
  array: ISelectOption[],
  key: string = 'value',
  separator: string = ',',
): string => {
  return _.chain(array).map(key).sortBy().join(separator).value()
}

// ==================== Order Utilities ====================
/**
 * Calculates the total price for an array of order items
 * @category Order
 * @param {IOrderItem[]} items - Array of order items
 * @returns {number} Total price
 */
const calculateTotalPrice = (items: IOrderItem[]) => {
  return items.reduce((sum, item) => sum + item.price * item.quantity, 0)
}

export {
  // Date utils
  parseDate,
  formatDateTime,
  localeMap,

  // Number & Price utils
  formatNumber,
  formatPrice,

  // Array utils
  transformToSortedString,
  // Order utils
  calculateTotalPrice,
}
