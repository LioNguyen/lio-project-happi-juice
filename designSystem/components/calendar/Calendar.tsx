import { Locale, vi } from 'date-fns/locale'
import { ChevronLeft, ChevronRight } from 'lucide-react'
import * as React from 'react'
import { DayPicker } from 'react-day-picker'

import { cn } from '@/shared/utils'
import { buttonVariants } from '@designSystem/components/button'

type CalendarProps = React.ComponentProps<typeof DayPicker> & {
  fontClassName?: string
  calendarLocale?: Locale
  customFormatCaption?: (date: Date) => string
  compact?: boolean
}

function Calendar({
  className,
  classNames,
  showOutsideDays = true,
  fontClassName,
  calendarLocale = vi,
  customFormatCaption,
  compact = false,
  ...props
}: CalendarProps) {
  const defaultFormatCaption = (date: Date) => {
    return `Tháng ${date.getMonth() + 1} năm ${date.getFullYear()}`
  }

  return (
    <DayPicker
      showOutsideDays={showOutsideDays}
      className={cn(compact ? 'p-1.5 w-full text-sm' : 'p-3 w-full', className)}
      locale={calendarLocale}
      formatters={{
        formatCaption: customFormatCaption || defaultFormatCaption,
      }}
      classNames={{
        months: cn(
          'flex flex-col sm:flex-row sm:space-x-4 w-full',
          compact ? 'space-y-2 sm:space-y-0' : 'space-y-4 sm:space-y-0',
        ),
        month: cn('w-full', compact ? 'space-y-2' : 'space-y-4'),
        caption: cn(
          'flex justify-center relative items-center',
          compact ? 'pt-0.5' : 'pt-1',
          fontClassName,
        ),
        caption_label: cn(
          compact ? 'text-xs font-medium' : 'text-sm font-medium',
          fontClassName,
        ),
        nav: 'space-x-1 flex items-center',
        nav_button: cn(
          buttonVariants({ variant: 'outline' }),
          compact ? 'h-5 w-5 p-0' : 'h-7 w-7 p-0',
          'bg-transparent opacity-50 hover:opacity-100',
        ),
        nav_button_previous: 'absolute left-1',
        nav_button_next: 'absolute right-1',
        table: 'w-full border-collapse',
        head_row: cn('flex justify-between', compact ? 'my-0.5' : 'mb-1'),
        head_cell: cn(
          'text-text-muted-foreground rounded-md font-normal',
          compact ? 'text-[0.7rem] w-7' : 'text-[0.8rem] w-9',
          fontClassName,
        ),
        row: cn('flex w-full justify-between', compact ? 'mt-1' : 'mt-2'),
        cell: cn(
          'relative p-0 text-center focus-within:relative focus-within:z-20 [&:has([aria-selected])]:bg-accent [&:has([aria-selected].day-outside)]:bg-accent/50 [&:has([aria-selected].day-range-end)]:rounded-r-md',
          compact ? 'text-xs' : 'text-sm',
          props.mode === 'range'
            ? '[&:has(>.day-range-end)]:rounded-r-md [&:has(>.day-range-start)]:rounded-l-md first:[&:has([aria-selected])]:rounded-l-md last:[&:has([aria-selected])]:rounded-r-md'
            : '[&:has([aria-selected])]:rounded-md',
        ),
        day: cn(
          buttonVariants({ variant: 'ghost' }),
          compact ? 'h-7 w-7 p-0 text-xs' : 'h-9 w-9 p-0 text-sm',
          'font-normal aria-selected:opacity-100',
          fontClassName,
        ),
        day_range_start: 'day-range-start',
        day_range_end: 'day-range-end',
        day_selected:
          'bg-background-secondary text-text-secondary hover:bg-background-secondary hover:text-text-secondary focus:bg-background-secondary focus:text-text-secondary',
        day_today: 'bg-accent text-accent-foreground',
        day_outside:
          'day-outside text-text-muted-foreground aria-selected:bg-accent/50 aria-selected:text-text-muted-foreground',
        day_disabled: 'text-text-muted-foreground opacity-50',
        day_range_middle:
          'aria-selected:bg-accent aria-selected:text-accent-foreground',
        day_hidden: 'invisible',
        ...classNames,
      }}
      components={{
        IconLeft: ({ className, ...props }) => (
          <ChevronLeft
            className={cn(compact ? 'h-3 w-3' : 'h-4 w-4', className)}
            {...props}
          />
        ),
        IconRight: ({ className, ...props }) => (
          <ChevronRight
            className={cn(compact ? 'h-3 w-3' : 'h-4 w-4', className)}
            {...props}
          />
        ),
      }}
      {...props}
    />
  )
}
Calendar.displayName = 'Calendar'

export { Calendar }
export type { CalendarProps }
