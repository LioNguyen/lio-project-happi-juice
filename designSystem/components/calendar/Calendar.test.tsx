import { ComponentPropsWithRef } from "react";
import { DayPicker } from "react-day-picker";
import { describe, expect, it, vi } from "vitest";

import { renderWithTheme } from "@/shared/utils";
import { Calendar } from "./Calendar";

// Mock DayPicker
vi.mock("react-day-picker", () => ({
  DayPicker: vi.fn(() => null),
}));

type CalendarPropsWithRef = ComponentPropsWithRef<typeof Calendar>;

const renderComponent = (props: Partial<CalendarPropsWithRef> = {}) => {
  return renderWithTheme(<Calendar {...props} />);
};

describe("Calendar Component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  // Snapshot Test
  it("should match snapshot with default props", () => {
    const { container } = renderComponent();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render DayPicker with default props", () => {
    renderComponent();

    expect(DayPicker).toHaveBeenCalledWith(
      expect.objectContaining({
        showOutsideDays: true,
        className: "p-3",
      }),
      expect.anything(),
    );
  });

  it("should merge custom className with default className", () => {
    renderComponent({ className: "custom-class" });

    expect(DayPicker).toHaveBeenCalledWith(
      expect.objectContaining({
        className: "p-3 custom-class",
      }),
      expect.anything(),
    );
  });

  it("should override showOutsideDays when provided", () => {
    renderComponent({ showOutsideDays: false });

    expect(DayPicker).toHaveBeenCalledWith(
      expect.objectContaining({
        showOutsideDays: false,
      }),
      expect.anything(),
    );
  });

  it("should pass correct classNames for default mode", () => {
    renderComponent();

    const dayPickerProps = vi.mocked(DayPicker).mock.calls[0][0];

    expect(dayPickerProps.classNames).toMatchObject({
      months: "flex flex-col sm:flex-row space-y-4 sm:space-x-4 sm:space-y-0",
      caption: "flex justify-center pt-1 relative items-center",
      cell: expect.stringContaining("[&:has([aria-selected])]:rounded-md"),
    });
  });

  it("should modify cell className when mode is range", () => {
    renderComponent({ mode: "range" });

    const dayPickerProps = vi.mocked(DayPicker).mock.calls[0][0];

    expect(dayPickerProps.classNames?.cell).toContain("[&:has(>.day-range-end)]:rounded-r-md");
    expect(dayPickerProps.classNames?.cell).toContain("[&:has(>.day-range-start)]:rounded-l-md");
  });

  it("should merge custom classNames with default classNames", () => {
    const customClassNames = {
      months: "custom-months",
      caption: "custom-caption",
    };

    renderComponent({ classNames: customClassNames });

    const dayPickerProps = vi.mocked(DayPicker).mock.calls[0][0];

    expect(dayPickerProps.classNames?.months).toBe("custom-months");
    expect(dayPickerProps.classNames?.caption).toBe("custom-caption");
    expect(dayPickerProps.classNames?.table).toBe("w-full border-collapse space-y-1");
  });

  it("should pass through additional props to DayPicker", () => {
    const additionalProps = {
      disabled: true,
      fromDate: new Date(),
      toDate: new Date(),
    };

    renderComponent(additionalProps);

    expect(DayPicker).toHaveBeenCalledWith(expect.objectContaining(additionalProps), expect.anything());
  });
});
