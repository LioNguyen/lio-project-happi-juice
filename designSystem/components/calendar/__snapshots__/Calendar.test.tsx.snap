// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Calendar Component > should match snapshot 1`] = `null`;

exports[`Calendar Component > should match snapshot with default props 1`] = `null`;
